### PR TITLE
Change PackageDescriptor.dependencies to DependencyDescriptors

### PR DESCRIPTION
**Pull request by murphm8**
This is for #106. 

I left some TODO in the code with questions. I think I caught all the usages in colcon_core. Anything using TaskContext.dependencies should be correct because that object is created in build.py/test.py where I have updated the usage of PackageDescriptor.dependencies.

I tested this with my updated colcon-ros in a simple ROS workspace. Do you have some test workspaces I can run this on to verify correctness? I don't have a python workspace to test against.

### CODE REVIEW

File: colcon_core/command.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def main(*, command_name='colcon', argv=None):
     """
     global colcon_logger
     # default log level
-    colcon_logger.setLevel(logging.WARN)
+    colcon_logger.setLevel(logging.INFO)
```
Comment (dirk-thomas) [inline colcon_core/command.py]: Please revert this unrelated change.

File: colcon_core/dependency_descriptor.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# Copyright 2016-2018 Dirk Thomas
+# Licensed under the Apache License, Version 2.0
+
+
+class DependencyDescriptor:
+    """A descriptor for a package dependency."""
+
```
Comment (dirk-thomas) [inline colcon_core/dependency_descriptor.py]: Please use `__slots__` to limit the members to a known set (`name`, `metadata`).

File: colcon_core/dependency_descriptor.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    def __init__(self, name, metadata=None):
```
Comment (dirk-thomas) [inline colcon_core/dependency_descriptor.py]: Force the keyword argument: `def __init__(self, name, *, metadata=None):`

File: colcon_core/dependency_descriptor.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        :param str name: Name of the declared dependency
+        :param dict metadata: Any metadata associated with
+        the dependency
+        """
+        self.name = name
+
+        if metadata is None:
+            self.metadata = {}
+        else:
+            self.metadata = metadata
```
Comment (dirk-thomas) [inline colcon_core/dependency_descriptor.py]: Nipick: shorten to `self.metadata = metadata if metadata is not None else {}`.

File: colcon_core/dependency_descriptor.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    def __eq__(self, other):  # noqa: D105
+        return self.name == other.name
+
+    def __lt__(self, other):  # noqa: D105
+        return self.name < other.name
+
+    def __hash__(self):  # noqa: D105
+        return hash(self.name)
```
Comment (dirk-thomas) [inline colcon_core/dependency_descriptor.py]: Generally this is dangerous since after an instance has been hashed the name could be changed which would result in a different hash. I don't have a better proposal though. Maybe prevent setting the `name` outside of the constructor in this case?

File: colcon_core/dependency_descriptor.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    def __str__(self):  # noqa: D105
+        return self.name
+
+    def __repr__(self):  # noqa: D105
+        return 'DependencyDescriptor({name})'.format(name=self.name)
```
Comment (dirk-thomas) [inline colcon_core/dependency_descriptor.py]: I would rather not define `__repr__` is not necessarily for anything?

File: colcon_core/environment/__init__.py [hunk 7]
```diff
@@ -126,8 +126,12 @@ def create_environment_scripts(
     # create file containing the runtime dependencies
     path = prefix_path / get_relative_package_index_path() / pkg.name
     path.parent.mkdir(parents=True, exist_ok=True)
+    dependencies = [dependency.name
+                    for dependency in pkg.dependencies.get('run', set())]
```
Comment (dirk-thomas) [inline colcon_core/environment/__init__.py]: Nitpick: shorten local variable name to: `dependencies = [d.name for d in pkg.dependencies.get('run', set())]`

In general I try to avoid aligned indentation. Maybe the new code fits in a single line. If not please wrap after the open parenthesis / bracket.

File: colcon_core/environment/__init__.py [hunk 8]
```diff
@@ -126,8 +126,12 @@ def create_environment_scripts(
...
+    # TODO: Is this supposed to be in topological order? I don't think the
+    # set is ordered?
```
Comment (dirk-thomas) [inline colcon_core/environment/__init__.py]: No, these dependencies are not ordered. Alphabetical only for deterministic results.

File: colcon_core/package_descriptor.py [hunk 9]
```diff
@@ -67,7 +67,8 @@ def get_dependencies(self, *, categories=None):
             categories = self.dependencies.keys()
         for category in sorted(categories):
             dependencies |= self.dependencies[category]
-        assert self.name not in dependencies, \
+        dependency_names = {dependency.name for dependency in dependencies}
```
Comment (dirk-thomas) [inline colcon_core/package_descriptor.py]: Same: `dependency_names = {d.name for d in dependencies}

File: colcon_core/package_identification/python.py [hunk 10]
```diff
@@ -100,5 +101,5 @@ def extract_dependencies(options):
             # remove environmental markers (separated by semicolons)
             # and version specifiers (separated by comparison operators)
             name = re.split(r';|<|>|<=|>=|==|!=', dep)[0].rstrip()
-            dependencies[dependency_type].add(name)
+            dependencies[dependency_type].add(DependencyDescriptor(name))
```
Comment (dirk-thomas) [inline colcon_core/package_identification/python.py]: The version constraints of the string should be extracted and added to the descriptor. This could be done in a separate patch if you prefer though.

File: colcon_core/topological_order.py [hunk 11]
```diff
@@ -39,14 +39,18 @@ def topological_order_packages(
         # remove dependencies on already ordered packages
         ordered_names = {descriptor.name for descriptor in ordered.keys()}
         for q in queued:
-            q.remaining_dependencies -= ordered_names
+            deps = set(filter(lambda dependency:
+                              dependency.name not in ordered_names,
+                              q.remaining_dependencies))
+            q.remaining_dependencies = deps
```
Comment (dirk-thomas) [inline colcon_core/topological_order.py]: Instead of modifying the algorithm (which doesn't need to know about versions) I would prefer changing the input to still only contain the names rather than full dependency descriptors (I think `rec_deps` above).

File: colcon_core/verb/test.py [hunk 12]
```diff
@@ -214,7 +214,9 @@ def _get_jobs(self, args, decorators, install_base):
 
             recursive_dependencies = OrderedDict()
             # for testing a package include itself in the environment
-            for dep_name in decorator.recursive_dependencies + [pkg.name]:
+            dependency_names = [dep.name
```
Comment (dirk-thomas) [inline colcon_core/verb/test.py]: Nitpick: please wrap after the bracket.

File: test/test_package_augmentation.py [hunk 13]
```diff
@@ -107,29 +108,34 @@ def test_update_descriptor():
     assert len(desc.metadata) == 0
 
     data = {
-        'build-dependencies': {'b1', 'b2'},
-        'test-dependencies': {'t1'},
+        'build-dependencies': {DependencyDescriptor('b1'),
```
Comment (dirk-thomas) [inline test/test_package_augmentation.py]: Nitpick: please wrap after the curly brace.

Same below.

File: test/test_package_descriptor.py [hunk 14]
```diff
@@ -33,15 +34,18 @@ def test_identifies_package():
...
-    d1.dependencies['run'].add('run-depend')
-    d1.dependencies['run'].add('depend')
-    assert d1.get_dependencies() == {'build-depend', 'run-depend', 'depend'}
-
-    d1.dependencies['test'].add('self')
+    d1.dependencies['build'].add(DependencyDescriptor('build-depend'))
+    d1.dependencies['build'].add(DependencyDescriptor('depend'))
+    d1.dependencies['run'].add(DependencyDescriptor('run-depend'))
+    d1.dependencies['run'].add(DependencyDescriptor('depend'))
+    # TODO: What if the run and build dependencies have different versions?
```
Comment (dirk-thomas) [inline test/test_package_descriptor.py]: I would say the more strict version constraint should be used when merging. In case of a conflict an `error` message should be logged.

Comment (murphm8) [reply test/test_package_descriptor.py]: Should this logic be on the implementer of the package identification extension point instead?
 
We could then say that in colcon-core there will be no conflicting dependencies.

File: colcon_core/environment/__init__.py (hunk 8 above)
Comment (dirk-thomas) [reply colcon_core/environment/__init__.py]: I think the comment it still valid as it is. The file contains "the runtime dependencies" and it is being used to "source all packages in topological order", It doesn't imply that the order in the file itself is ordered.

File: colcon_core/dependency_descriptor.py (hunk 5 above)
Comment (murphm8) [reply colcon_core/dependency_descriptor.py]: Converted name to be a read-only property. 

The other issue I see here is that two DependencyDescriptor with different metadata but the same name will be considered equal. Should __eq__ contain the metadata dictionary too? That would require everything stored in metadata to be comparable, is that okay?

File: colcon_core/environment/__init__.py [hunk 15]
```diff
@@ -126,8 +126,12 @@ def create_environment_scripts(
...
     path.write_text(
-        os.pathsep.join(sorted(pkg.dependencies.get('run', set()))))
+        os.pathsep.join(sorted(dependencies)))
```
Comment (murphm8) [inline colcon_core/environment/__init__.py]: This functions docstring states:     

:returns: iterable of the generated hook paths
:rtype: Iterable

But there is no return statement, is that an issue?

File: colcon_core/package_identification/python.py (hunk 10 above)
Comment (murphm8) [reply colcon_core/package_identification/python.py]: Yea, I'll do that in a separate commit.

File: colcon_core/dependency_descriptor.py (hunk 5 above)
Comment (dirk-thomas) [reply colcon_core/dependency_descriptor.py]: Another option would be to just not override `__eq__` and wait until there is a use case for it.

File: colcon_core/environment/__init__.py (hunk 15 above)
Comment (dirk-thomas) [reply colcon_core/environment/__init__.py]: I think the docstring is just wrong. Please address this in a separate PR (otherwise I can do that too if you prefer).

File: colcon_core/topological_order.py (hunk 11 above)
Comment (murphm8) [reply colcon_core/topological_order.py]: That's much simpler! Could you explain the difference between PackageDecorator and PackageDescriptor? I'm concerned PackageDecorator.recursive_dependencies is a different type than PackageDescriptor.dependencies if I simplify this.

Comment (dirk-thomas) [reply colcon_core/topological_order.py]: https://github.com/colcon/colcon-core/blob/bdba48223fbd73448caf424a934faaec35e8b4ee/colcon_core/package_descriptor.py#L8-L22

https://github.com/colcon/colcon-core/blob/bdba48223fbd73448caf424a934faaec35e8b4ee/colcon_core/package_decorator.py#L16-L24

**Comment by murphm8** [comment]
Hey, I'm going to be on vacation the rest of the week but will pick this up on Monday. Thanks for all your feedback!

File: colcon_core/dependency_descriptor.py (hunk 5 above)
Comment (murphm8) [reply colcon_core/dependency_descriptor.py]: Overriding `__eq__` is nice for unit testing, not sure if that is enough of a reason? 

Should I just make `__eq__` use `(self.name, self.metadata)`, and have `__hash__` just use `self.name`?

Comment (murphm8) [reply colcon_core/dependency_descriptor.py]: It's needed in a couple different tests where we check dependencies. It would also be needed by any consumers of colcon-core that want to test using DependencyDescriptor.

I think keeping `__eq__` would be simpler.

Comment (dirk-thomas) [reply colcon_core/dependency_descriptor.py]: > It would also be needed by any consumers of colcon-core that want to test using DependencyDescriptor.

Do we know what potential consumers want to test for? If not, I would rather leave it out for now until that is the case. We can always add it later when there is a concrete need for it.

Comment (murphm8) [reply colcon_core/dependency_descriptor.py]: Looking into what happens when I remove `__eq__`, we get duplicates in the set. If I remove `__eq__` then I see unit tests where the same dependency is listed twice in the set.

    def test_get_dependencies():
        d1 = PackageDescriptor('/some/path')
        d1.name = 'self'
        d1.dependencies['build'].add(DependencyDescriptor('build-depend'))
        d1.dependencies['build'].add(DependencyDescriptor('depend'))
        d1.dependencies['run'].add(DependencyDescriptor('run-depend'))
        d1.dependencies['run'].add(DependencyDescriptor('depend'))
    >       assert check_dependencies(d1.get_dependencies(),
                                  ['build-depend', 'run-depend', 'depend'])
    E       AssertionError: assert False
    E        +  where False = check_dependencies({{_name: build-depend, metadata: {}}, {_name:     depend, metadata: {}}, {_name: depend, metadata: {}}, {_name: run-depend, metadata: {}}},     ['build-depend', 'run-depend', 'depend'])
    E        +    where {{_name: build-depend, metadata: {}}, {_name: depend, metadata: {}},     {_name: depend, metadata: {}}, {_name: run-depend, metadata: {}}} = <bound method PackageDescriptor.get_dependencies of <colcon_core.package_descriptor.PackageDescriptor object at 0x1074a3db0>>()
    E        +      where <bound method PackageDescriptor.get_dependencies of <colcon_core.package_descriptor.PackageDescriptor object at 0x1074a3db0>> = <colcon_core.package_descriptor.PackageDescriptor object at 0x1074a3db0>.get_dependencies

Above we can see that `d1.get_dependencies()` returns a set containing the "depend" dependency twice. If I define `__eq__` as `(this.name, this.metadata) == (other.name, other.metadata)` then there are no duplicates. 

It's my understanding that an object with a matching hash is not necessarily equal. So `obj1.hash == obj2.hash` doesn't imply `obj1 == obj2`, but `obj1 == obj2` implies `obj1.hash == obj2.hash`. This means that the set semantics will keep both copies of the object if `__eq__` is not defined.

Comment (dirk-thomas) [reply colcon_core/dependency_descriptor.py]: > but `obj1 == obj2` implies `obj1.hash == obj2.hash`.

Indeed, looking at the Python documentation of [__hash__](https://docs.python.org/3/reference/datamodel.html#object.__hash__):

    "The only required property is that objects which compare equal have the same hash value"

> Above we can see that d1.get_dependencies() returns a set containing the "depend" dependency twice.

The return value of `get_dependencies()` *should* contain "duplicates" because two dependencies with the same name do not necessarily have the same version constraints.

So if you store dependencies in a `set` (as this function does in the return value) it *must not* de-duplicate dependencies solely based on their name since the metadata might be different. As a direct consequence the hash can't be computed based on the name alone.

The question if dependencies which have the same name and metadata should be de-duplicated is a separate one: the function itself could (not must) de-duplicate dependencies which are equal in all the properties. That could either happen within the logic of the function or if `__hash__` is defined across all properties (instead of relying on the object id). And again in this case my previous concern applies again: exposing the metadata which allows callers to manipulate the data *after* a dependency has been hashed is dangerous.

Therefore I am still advocating to neither define `__hash__` nor `__eq__` since I don't see the need for them.

File: colcon_core/dependency_descriptor.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+        self._name = name
+        self.metadata = metadata if metadata is not None else {}
+
+    @property
+    def name(self):
+        """Name of the dependency."""
+        return self._name
+
+    def __lt__(self, other):  # noqa: D105
+        return self.name < other.name
```
Comment (dirk-thomas) [inline colcon_core/dependency_descriptor.py]: Why is this one necessary?

File: colcon_core/package_descriptor.py [hunk 17]
```diff
@@ -87,17 +88,18 @@ def get_recursive_dependencies(
           categories
         :param Iterable[str] recursive_categories: The names of the recursive
           categories
-        :returns: The names of the recursive dependencies
+        :returns: DependencyDescriptor objects of recursive dependencies
```
Comment (dirk-thomas) [inline colcon_core/package_descriptor.py]: Please mention the type of the return value as part of the `rtype`:

```
:returns: The recursive dependencies
:rtype: set[DependencyDescriptor]
```

File: colcon_core/dependency_descriptor.py (hunk 16 above)
Comment (murphm8) [reply colcon_core/dependency_descriptor.py]: Removed

File: colcon_core/package_descriptor.py (hunk 17 above)
Comment (murphm8) [reply colcon_core/package_descriptor.py]: Updated both docstrings.

**Comment by murphm8** [comment]
To roll this out do we need to make a pre-change to all the packages that would break? Something like lock their colcon-core dependency to <= 0.3.9. Then make the change and update them all to depend on >= 0.3.10?

**Comment by dirk-thomas** [comment]
> Something like lock their colcon-core dependency

That wouldn't help since we can't guarantee that user will actually get that interim version. They might just stick with the version the already have.

**Comment by murphm8** [comment]
How have you handled breaking changes like this in the past? Is there anything I can do?

**Comment by dirk-thomas** [comment]
> How have you handled breaking changes like this in the past?

Since the versions are still 0.x it is technically fine to break anything in a minor release. That being said it is obviously not desired when it can be avoided. Also looking into the future it would be good to figure out the right approach how to handle this (if we would be post 1.0 and don't want to level the major version).

> Is there anything I can do?

I have to think about it (when I find a bit more time) if there is a way to do it without breakage (e.g. subclassing from `str`) or what should be changed so that this is catchable gracefully in the future.

**Comment by murphm8** [comment]
I pushed a new commit that adds backwards compatibility with existing extensions that treat PackageDescriptor.dependencies as a list of strings. I've  updated the tests to include both DependencyDescriptor and str dependencies in the dependencies list.

**Comment by dirk-thomas** [comment]
I don't see how the updated patch achieves backward compatibility. Existing code is expecting a `str` and doesn't know about / use the new function `dependency_name`. I haven't tried the patch but don't see how that would work in a backward compatible fashion. Can you please elaborate how you tested backward compatibility (e.g. using this branch of `colcon-core` in combination with existing released extensions).

**Comment by murphm8** [comment]
You're correct. This will break existing consumers that depend on PackageDescriptor.dependencies being a list of strings. 

The only way I can think to support both would be to create a new attribute on PackageDescriptor for DependencyDescriptors. Then a setter and getter for PackageDescriptor.dependencies. 

The setter would take a list of strings and create the PackageDescriptor.dependeny_descriptors list. 

The getter would return [d.name for d in descriptor.dependency_descriptors]

What do you think of that approach? Would you be okay with a new attribute on PackageDescriptor?

**Comment by murphm8** [comment]
I'm looking into subclassing str. The one issue I'd have with that is then anywhere that wants to access metadata on a DependencyDescriptor needs to first verify isInstance(d, DependencyDescriptor). Do you think that's okay?

**Comment by murphm8** [comment]
With this change I've verified that the colcon-package-information (colcon info) in PyPI works with the latest colcon-core and colcon-ros.

**Comment by dirk-thomas** [comment]
I reviewed the current state of the patch and a few things looked "weird" to me - or at least they were unclear why they are necessary:

* inheriting from `str` but still having a `_name` slot
* inheriting from `str` but still having `__slots__`
* (the need to) override the `__eq__` / `__hash__` methods
* the need for `dependency_name` - the goal is that existing code not using that function are still compatible
* the "logic" changes in `get_dependencies` / `get_recursive_dependencies` to use the name rather then the descriptor

Instead of iterating on this patch I tried locally if we can't get away with a much smaller patch. I prototyped #116 and it seems to work for me. Can you please give that patch a try and comment if it works for you too.

**Comment by murphm8** [comment]
I agree your change is much cleaner and more concise. I found I needed to override __eq__ and __hash__ to get DependencyDescriptor to work correctly as a `str` subclass. 

The `dependency_name` function could go away. The changes in `get_dependencies` and `get_recursive_dependencies` would also not be necessary if DependencyDescriptor can be treated as a string.

**Comment by dirk-thomas** [comment]
> The changes in get_dependencies and get_recursive_dependencies would also not be necessary if DependencyDescriptor can be treated as a string.

I included these changes so that "consumers" of the functions get a consistent return value and don't have to deal with two potential types.

**Comment by dirk-thomas** [comment]
> I found I needed to override __eq__ and __hash__ to get DependencyDescriptor to work correctly as a str subclass.

I didn't see a need for it. Can you double check if the current state works for you (without these methods overridden)?

**Comment by murphm8** [comment]
> I didn't see a need for it. Can you double check if the current state works for you (without these methods overridden)?

Posted on #116. Your change works for me.

> I included these changes so that "consumers" of the functions get a consistent return value and don't have to deal with two potential types.

I agree doing this will make it easier for consumers of those apis.

**Comment by dirk-thomas** [comment]
Closing in favor of #116.
